Hit this in CI on PR #654 — run [24318628083](https://github.com/gastownhall/gascity/actions/runs/24318628083/job/71000619475):

```
--- FAIL: TestReconcileSessionBeads_StartsIdleDrainAfterGrace (0.00s)
    session_sleep_test.go:231: idle probe for gc-1 not ready after waitIdleProbes returned
```

This is the intra-tick race, not the original wall-clock polling bug (#497/#510). The WaitGroup completes but the probe state isn't visible yet — exactly what this PR fixes.

--- a/cmd/gc/session_reconciler.go
+++ b/cmd/gc/session_reconciler.go
@@ -911,9 +911,7 @@ func launchIdleProbes(
 		if name == "" || probe == nil {
 			continue
 		}
-		dt.beginIdleProbe()
 		go func(beadID, sessionName string, probe *idleProbeState) {
-			defer dt.doneIdleProbe()
 			err := wp.WaitForIdle(ctx, sessionName, idleSleepProbeTimeout)
 			dt.finishIdleProbe(beadID, probe, err == nil, clk.Now().UTC())
 		}(target.session.ID, name, probe)

--- a/cmd/gc/session_reconciler_test.go
+++ b/cmd/gc/session_reconciler_test.go
@@ -1175,8 +1175,11 @@ func TestReconcileSessionBeads_PreservedRunningNamedSessionStillIdleDrains(t *te
 		t.Fatalf("start session: %v", err)
 	}
 	env.sp.WaitForIdleErrors[sessionName] = nil
+	idleGate := make(chan struct{}) // see waitForIdleProbeReady godoc
+	env.sp.WaitForIdleGates[sessionName] = idleGate
 
 	env.reconcile([]beads.Bead{session})
+	close(idleGate)
 	waitForIdleProbeReady(t, env.dt, session.ID)
 	env.reconcile([]beads.Bead{session})
 

--- a/cmd/gc/session_sleep_test.go
+++ b/cmd/gc/session_sleep_test.go
@@ -220,6 +220,8 @@ func TestReconcileSessionBeads_StartsIdleDrainAfterGrace(t *testing.T) {
 	session.Metadata["last_woke_at"] = ts
 	session.Metadata["detached_at"] = ts
 	env.sp.WaitForIdleErrors["worker"] = nil
+	idleGate := make(chan struct{}) // see waitForIdleProbeReady godoc
+	env.sp.WaitForIdleGates["worker"] = idleGate
 
 	poolDesired := map[string]int{"worker": 1}
 	cfgNames := configuredSessionNames(env.cfg, "", env.store)
@@ -228,6 +230,7 @@ func TestReconcileSessionBeads_StartsIdleDrainAfterGrace(t *testing.T) {
 		env.store, nil, nil, nil, env.dt, poolDesired, false, nil, "",
 		nil, env.clk, env.rec, 0, 0, &env.stdout, &env.stderr,
 	)
+	close(idleGate)
 	waitForIdleProbeReady(t, env.dt, session.ID)
 	reconcileSessionBeads(
 		context.Background(), []beads.Bead{session}, env.desiredState, cfgNames, env.cfg, env.sp,
@@ -984,22 +987,26 @@ func TestAdvanceSessionDrainsWithSessions_UsesProvidedWakeEvaluations(t *testing
 	}
 }
 
+// waitForIdleProbeReady polls the drain tracker until the idle probe for
+// beadID is registered and marked ready, or until the deadline expires.
+//
+// Callers must ensure the probe goroutine does not complete within the
+// same reconcile tick that started it — otherwise shouldBeginIdleDrain
+// observes probe.ready=true and its deferred clearIdleProbe removes the
+// probe from the map before this helper can see it. The standard way to
+// enforce that ordering is to set env.sp.WaitForIdleGates[sessionName]
+// to a channel, run the tick (probe goroutine blocks on the gate), then
+// close the gate and call this helper. Without the gate, tests race
+// against goroutine scheduling and flake under -race.
 func waitForIdleProbeReady(t *testing.T, dt *drainTracker, beadID string) {
 	t.Helper()
 
-	done := make(chan struct{})
-	go func() {
-		dt.waitIdleProbes()
-		close(done)
-	}()
-
-	select {
-	case <-done:
-	case <-time.After(10 * time.Second):
-		t.Fatalf("idle probe for %s did not complete within 10s", beadID)
+	deadline := time.Now().Add(10 * time.Second)
+	for time.Now().Before(deadline) {
+		if probe, ok := dt.idleProbe(beadID); ok && probe.ready {
+			return
+		}
+		time.Sleep(25 * time.Millisecond)
 	}
-
-	if probe, ok := dt.idleProbe(beadID); !ok || !probe.ready {
-		t.Fatalf("idle probe for %s not ready after waitIdleProbes returned", beadID)
-	}
+	t.Fatalf("idle probe for %s not ready within 10s", beadID)
 }

--- a/cmd/gc/session_types.go
+++ b/cmd/gc/session_types.go
@@ -71,7 +71,6 @@ type drainTracker struct {
 	drains          map[string]*drainState     // session bead ID -> drain state
 	idleProbes      map[string]*idleProbeState // session bead ID -> async idle probe
 	idleProbeCursor int
-	idleProbeWG     sync.WaitGroup
 }
 
 func newDrainTracker() *drainTracker {
@@ -176,27 +175,6 @@ func (dt *drainTracker) clearIdleProbe(beadID string) {
 	dt.mu.Lock()
 	defer dt.mu.Unlock()
 	delete(dt.idleProbes, beadID)
-}
-
-func (dt *drainTracker) beginIdleProbe() {
-	if dt == nil {
-		return
-	}
-	dt.idleProbeWG.Add(1)
-}
-
-func (dt *drainTracker) doneIdleProbe() {
-	if dt == nil {
-		return
-	}
-	dt.idleProbeWG.Done()
-}
-
-func (dt *drainTracker) waitIdleProbes() {
-	if dt == nil {
-		return
-	}
-	dt.idleProbeWG.Wait()
 }
 
 // Reconciler tuning defaults.

--- a/cmd/gc/session_types_test.go
+++ b/cmd/gc/session_types_test.go
@@ -96,36 +96,6 @@ func TestDrainTracker(t *testing.T) {
 	}
 }
 
-func TestDrainTracker_WaitIdleProbes(t *testing.T) {
-	dt := newDrainTracker()
-	release := make(chan struct{})
-	done := make(chan struct{})
-
-	dt.beginIdleProbe()
-	go func() {
-		defer dt.doneIdleProbe()
-		<-release
-	}()
-	go func() {
-		dt.waitIdleProbes()
-		close(done)
-	}()
-
-	select {
-	case <-done:
-		t.Fatal("waitIdleProbes returned before probe completion")
-	case <-time.After(20 * time.Millisecond):
-	}
-
-	close(release)
-
-	select {
-	case <-done:
-	case <-time.After(time.Second):
-		t.Fatal("waitIdleProbes did not unblock after probe completion")
-	}
-}
-
 func TestExecSpec_ZeroValue(t *testing.T) {
 	var spec ExecSpec
 	if spec.Path != "" || spec.WorkDir != "" {

--- a/internal/runtime/fake.go
+++ b/internal/runtime/fake.go
@@ -411,9 +411,9 @@ func (f *Fake) ClearScrollback(name string) error {
 
 // WaitForIdle records the call and returns the configured result. When
 // WaitForIdleGates[name] is set, the method releases f.mu and blocks on
-// <-gate before returning, giving tests deterministic control over when
-// the call completes.
-func (f *Fake) WaitForIdle(_ context.Context, name string, _ time.Duration) error {
+// the gate (or ctx cancellation) before returning, giving tests
+// deterministic control over when the call completes.
+func (f *Fake) WaitForIdle(ctx context.Context, name string, _ time.Duration) error {
 	f.mu.Lock()
 	f.Calls = append(f.Calls, Call{Method: "WaitForIdle", Name: name})
 	if f.broken {
@@ -428,7 +428,11 @@ func (f *Fake) WaitForIdle(_ context.Context, name string, _ time.Duration) erro
 	gate := f.WaitForIdleGates[name]
 	f.mu.Unlock()
 	if gate != nil {
-		<-gate
+		select {
+		case <-gate:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
 	}
 	return err
 }

--- a/internal/runtime/fake.go
+++ b/internal/runtime/fake.go
@@ -28,6 +28,12 @@ type Fake struct {
 	Responses            map[string][]InteractionResponse
 	SleepCapabilityValue SessionSleepCapability
 	WaitForIdleErrors    map[string]error
+	// WaitForIdleGates blocks WaitForIdle on a per-name channel until the
+	// caller closes it. A nil or absent entry returns the configured
+	// WaitForIdleErrors value immediately. The gate is read under f.mu
+	// and the lock is released before the block, so other Fake methods
+	// remain callable while a probe is gated.
+	WaitForIdleGates map[string]chan struct{}
 }
 
 // Call records a single method invocation on [Fake].
@@ -57,6 +63,7 @@ func NewFake() *Fake {
 		Responses:            make(map[string][]InteractionResponse),
 		SleepCapabilityValue: SessionSleepCapabilityFull,
 		WaitForIdleErrors:    make(map[string]error),
+		WaitForIdleGates:     make(map[string]chan struct{}),
 	}
 }
 
@@ -74,6 +81,7 @@ func NewFailFake() *Fake {
 		Responses:            make(map[string][]InteractionResponse),
 		SleepCapabilityValue: SessionSleepCapabilityFull,
 		WaitForIdleErrors:    make(map[string]error),
+		WaitForIdleGates:     make(map[string]chan struct{}),
 		broken:               true,
 	}
 }
@@ -401,17 +409,26 @@ func (f *Fake) ClearScrollback(name string) error {
 	return nil
 }
 
-// WaitForIdle records the call and returns the configured result.
+// WaitForIdle records the call and returns the configured result. When
+// WaitForIdleGates[name] is set, the method releases f.mu and blocks on
+// <-gate before returning, giving tests deterministic control over when
+// the call completes.
 func (f *Fake) WaitForIdle(_ context.Context, name string, _ time.Duration) error {
 	f.mu.Lock()
-	defer f.mu.Unlock()
 	f.Calls = append(f.Calls, Call{Method: "WaitForIdle", Name: name})
 	if f.broken {
+		f.mu.Unlock()
 		return fmt.Errorf("session unavailable")
 	}
 	err, ok := f.WaitForIdleErrors[name]
 	if !ok {
+		f.mu.Unlock()
 		return ErrInteractionUnsupported
+	}
+	gate := f.WaitForIdleGates[name]
+	f.mu.Unlock()
+	if gate != nil {
+		<-gate
 	}
 	return err
 }

--- a/internal/runtime/fake_test.go
+++ b/internal/runtime/fake_test.go
@@ -2,7 +2,9 @@ package runtime
 
 import (
 	"context"
+	"errors"
 	"testing"
+	"time"
 )
 
 // Compile-time check: Fake implements Provider.
@@ -403,4 +405,85 @@ func TestFlattenText_Empty(t *testing.T) {
 	if got := FlattenText([]ContentBlock{{Type: "text"}}); got != "" {
 		t.Errorf("FlattenText(empty text) = %q, want empty", got)
 	}
+}
+
+func TestFakeWaitForIdleGate_BlocksUntilClosed(t *testing.T) {
+	f := NewFake()
+	f.WaitForIdleErrors["s1"] = nil
+	gate := make(chan struct{})
+	f.WaitForIdleGates["s1"] = gate
+
+	done := make(chan error, 1)
+	go func() {
+		done <- f.WaitForIdle(context.Background(), "s1", time.Second)
+	}()
+
+	select {
+	case <-done:
+		t.Fatal("WaitForIdle returned before gate closed")
+	case <-time.After(50 * time.Millisecond):
+	}
+
+	close(gate)
+
+	select {
+	case err := <-done:
+		if err != nil {
+			t.Fatalf("WaitForIdle returned error: %v", err)
+		}
+	case <-time.After(time.Second):
+		t.Fatal("WaitForIdle did not return after gate closed")
+	}
+}
+
+func TestFakeWaitForIdleGate_RespectsContextCancel(t *testing.T) {
+	f := NewFake()
+	f.WaitForIdleErrors["s1"] = nil
+	f.WaitForIdleGates["s1"] = make(chan struct{}) // never closed
+
+	ctx, cancel := context.WithCancel(context.Background())
+	done := make(chan error, 1)
+	go func() {
+		done <- f.WaitForIdle(ctx, "s1", time.Second)
+	}()
+
+	select {
+	case <-done:
+		t.Fatal("WaitForIdle returned before cancel")
+	case <-time.After(50 * time.Millisecond):
+	}
+
+	cancel()
+
+	select {
+	case err := <-done:
+		if !errors.Is(err, context.Canceled) {
+			t.Fatalf("WaitForIdle error = %v, want context.Canceled", err)
+		}
+	case <-time.After(time.Second):
+		t.Fatal("WaitForIdle did not return after context cancel")
+	}
+}
+
+func TestFakeWaitForIdleGate_MuReleasedWhileBlocked(t *testing.T) {
+	f := NewFake()
+	_ = f.Start(context.Background(), "s1", Config{WorkDir: "/tmp"})
+	f.WaitForIdleErrors["s1"] = nil
+	gate := make(chan struct{})
+	f.WaitForIdleGates["s1"] = gate
+
+	// Start a gated WaitForIdle in the background.
+	go func() {
+		_ = f.WaitForIdle(context.Background(), "s1", time.Second)
+	}()
+
+	// Give the goroutine time to acquire and release the lock.
+	time.Sleep(20 * time.Millisecond)
+
+	// Other Fake operations must not deadlock while the gate is held.
+	if !f.IsRunning("s1") {
+		t.Fatal("IsRunning returned false while gate is held")
+	}
+
+	close(gate)
 }

